service/src/config.rs

```diff
@@ -1,61 +1,312 @@
+		let mu_ra_port = "99";
+		let expected_worker_ip = "127.0.0.1";
+
+		let mut args = ArgMatches::default();
+		args.args = HashMap::from([
+			("mu-ra-port", Default::default()),
+			("untrusted-worker-port", Default::default()),
+			("trusted-worker-port", Default::default()),
+		]);
+		// Workaround because MatchedArg is private.
+		args.args.get_mut("mu-ra-port").unwrap().vals = vec![mu_ra_port.into()];
+		args.args.get_mut("untrusted-worker-port").unwrap().vals = vec![untrusted_port.into()];
+		args.args.get_mut("trusted-worker-port").unwrap().vals = vec![trusted_port.into()];
+
+		let config = Config::from(&args);
+
+		assert_eq!(
+			config.trusted_worker_url_external(),
+			format!("wss://{}:{}", expected_worker_ip, trusted_port)
+		);
+		assert_eq!(
+			config.untrusted_worker_url_external(),
+			format!("ws://{}:{}", expected_worker_ip, untrusted_port)
+		);
+		assert_eq!(config.mu_ra_url_external(), format!("{}:{}", expected_worker_ip, mu_ra_port));
+	}
+
+	#[test]
+	fn external_addresses_are_returned_correctly_if_given() {
+		let trusted_ext_addr = "wss://1.1.1.2:700";
+		let untrusted_ext_addr = "ws://1.723.3.1:11";
+		let mu_ra_ext_addr = "1.1.3.1:1000";
+
+		let mut args = ArgMatches::default();
+		args.args = HashMap::from([
+			("trusted-external-address", Default::default()),
+			("untrusted-external-address", Default::default()),
+			("mu-ra-external-address", Default::default()),
+		]);
+		// Workaround because MatchedArg is private.
+		args.args.get_mut("trusted-external-address").unwrap().vals = vec![trusted_ext_addr.into()];
+		args.args.get_mut("untrusted-external-address").unwrap().vals =
+			vec![untrusted_ext_addr.into()];
+		args.args.get_mut("mu-ra-external-address").unwrap().vals = vec![mu_ra_ext_addr.into()];
+
+		let config = Config::from(&args);
+
+		assert_eq!(config.trusted_worker_url_external(), trusted_ext_addr);
+		assert_eq!(config.untrusted_worker_url_external(), untrusted_ext_addr);
+		assert_eq!(config.mu_ra_url_external(), mu_ra_ext_addr);
+	}
+
+	#[test]
+	fn add_port_if_necessary_does_not_add_port_if_given() {
+		let url = "ws://hello:4000";
+		let port = "0";
+
+		let resulting_url = add_port_if_necessary(url, port);
+
+		assert_eq!(resulting_url, url);
+	}
+
+	#[test]
+	fn add_port_if_necessary_adds_port_if_not_given() {
+		let url = "wss://hello";
+		let port = "0";
+
+		let resulting_url = add_port_if_necessary(url, port);
+
+		assert_eq!(resulting_url, format!("{}:{}", url, port));
+	}
+
+	#[test]
+	fn without_ws_add_port_if_necessary_does_not_add_port_if_given_without_ws() {
+		let url = "hello:10001";
+		let port = "012";
+
+		let resulting_url = add_port_if_necessary(url, port);
+
+		assert_eq!(resulting_url, url);
+	}
+
+	#[test]
+	fn add_port_if_necessary_adds_port_if_not_given_without_ws() {
+		let url = "hello_world";
+		let port = "10";
+
+		let resulting_url = add_port_if_necessary(url, port);
+
+		assert_eq!(resulting_url, format!("{}:{}", url, port));
+	}
```

Comment:
Wow, nice testing.

Just a comment, your test-names are sometimes too verbose that I don't really get what they are testing.

Reply:
True enough, I also wasn't happy. Let me think about some better names.

Reply:
Updated - but making them "short" was not really possible. Hope it's better understandable now, either way.